xds/internal/balancer/cdsbalancer/cdsbalancer.go

```diff
@@ -145,15 +171,191 @@ type watchUpdate struct {
+	cancelWatch    func()                // Cluster watch cancel func.
+	edsLB          balancer.Balancer     // EDS child policy.
 	clusterToWatch string
 	logger         *grpclog.PrefixLogger
 	closed         *grpcsync.Event
+
+	// The certificate providers are cached here to that they can be closed when
+	// a new provider is to be created.
+	cachedRoot     certprovider.Provider
+	cachedIdentity certprovider.Provider
+	xdsHI          *xds.HandshakeInfo
+	xdsCredsInUse  bool
+}
+
+// handleClientConnUpdate handles a ClientConnUpdate received from gRPC. Good
+// updates lead to registration of a CDS watch. Updates with error lead to
+// cancellation of existing watch and propagation of the same error to the
+// edsBalancer.
+func (b *cdsBalancer) handleClientConnUpdate(update *ccUpdate) {
+	// We first handle errors, if any, and then proceed with handling the
+	// update, only if the status quo has changed.
+	if err := update.err; err != nil {
+		b.handleErrorFromUpdate(err, true)
+	}
+	if b.xdsClient == update.client && b.clusterToWatch == update.clusterName {
+		return
+	}
+	if update.client != nil {
+		// Since the cdsBalancer doesn't own the xdsClient object, we don't have
+		// to bother about closing the old client here, but we still need to
+		// cancel the watch on the old client.
+		b.cancelWatch()
+		b.xdsClient = update.client
+	}
+	if update.clusterName != "" {
+		cancelWatch := b.xdsClient.WatchCluster(update.clusterName, b.handleClusterUpdate)
+		b.logger.Infof("Watch started on resource name %v with xds-client %p", update.clusterName, b.xdsClient)
+		b.cancelWatch = func() {
+			cancelWatch()
+			b.logger.Infof("Watch cancelled on resource name %v with xds-client %p", update.clusterName, b.xdsClient)
+		}
+		b.clusterToWatch = update.clusterName
+	}
+}
+
+// handleSecurityConfig processes the security configuration received from the
+// management server, creates appropriate certificate provider plugins, and
+// updates the HandhakeInfo which is added as an address attribute in
+// NewSubConn() calls.
+func (b *cdsBalancer) handleSecurityConfig(config *xdsclient.SecurityConfig) error {
+	// If xdsCredentials are not in use, i.e, the user did not want to get
+	// security configuration from an xDS server, we should not be acting on the
+	// received security config here. Doing so poses a security threat.
+	if !b.xdsCredsInUse {
+		return nil
+	}
+	// Security config being nil is a valid case where the management server has
+	// not sent any security configuration. The xdsCredentials implementation
+	// handles this by delegating to its fallback credentials.
+	if config == nil {
+		return nil
+	}
+
+	cpc := b.xdsClient.CertProviderConfigs()
+	if cpc == nil {
+		// Bootstrap did not find any certificate provider configs, but the user
+		// has specified xdsCredentials and the management server has sent down
+		// security configuration.
+		return errors.New("xds: certificate_providers config missing in bootstrap file")
```

Comment:
Should this failure happen when the xds creds are instantiated?  Or is it reasonable to use xds creds without being ready to handle them if they come?

Reply:
Would be hard to force this failure when the xdsCreds are instantiated. Either xdsCreds will have to read the bootstrap file, or the bootstrap processing code needs to know that xdsCreds are in use.

Also, there is one use-case where things can work normally when using xdsCreds with an empty `certificate_providers` config. This is the case where the control plane does not send down any security config in the `Cluster` resource. Maybe this could be useful as part of a transition from using static certs to moving to xds creds.

Reply:
> Also, there is one use-case where things can work normally

Why would they be using xds credentials if they don't have any `certificate_providers`?  This is a failure waiting to happen when the control plane changes.  It might be worth discussing this scenario cross-language, and whether we want to allow it or fail channel creation when it happens.

Reply:
Initiated a discussion. Thanks.

Reply:
It looks like the other languages also haven't handled this cleanly as of now, but it does look like something that can be handled without too much trouble.

- The `xdsResolver` is passed the credentials configured on the channel.
- The `xdsResolver` is the one which will create the `xdsClient`, and the bootstrap processing happens as part of that.
- So, the resolver can check if `xdsCreds` have been configured (using the same `UseXDS` function that the `cdsBalancer` uses), and in that case if the `certificate_providers` config is empty, it can fail `resolver.Build()` which will fail `grpc.Dial()`. 

Let me know if this feel ok to you.

Reply:
I can see this going either way: it could fail anyway if they haven't configured the *proper* credentials, but if they haven't configured *any* credentials then it's guaranteed to fail unless it's not needed.  Why would you use xds credentials if they aren't needed?  Is there a good reason?  Maybe to future-proof so all you need to do is update the bootstrap file and not recompile?  If that's a reasonable use-case then we shouldn't fail in this case.

I think we need a consensus among languages before doing anything here.

Reply:
Had a discussion with other security folks, and we think that we should not fail when `XdsCredentials` are specified by the user, but there is no `certificate_providers` config in the bootstrap file. The reasoning behind this is that the `CertProviderInstance` field in the `TlsContext` received from the control plane is just one way of specifying the security configuration. There are other ways of specifying the security config through the `TlsContext`. So just because there is no config in the bootstrap file doesn't mean that we will not be able to correctly provision security.